### DISCUSSION TITLE
Is there a way to get esp-mbedtls use PSRAM?

### DISCUSSION THREAD

**Discussion by yanshay** [esp-rs/esp-mbedtls#55] [Q&A] [answered]
I need to free memory, is it possible to use PSRAM instead of the standard heap for esp-mbedtls?

**Comment by bjoernQ** [comment]
For now, `esp-mbedtls` is using `malloc` from `esp-wifi` and `esp-wifi` currently needs to only allocate from internal RAM (we might lift that requirement some day)

In general it should be fine if `esp-mbedtls` would use PSRAM - at least for the software crypto. For using crypto accelerators with DMA things are a bit different (some chips can do DMA from/to PSRAM but have some special requirements then)

**Reply by yanshay** [reply]
I see in the code that allocations are done using an extern "C" `calloc` function.
So if I somehow get `calloc` to be the standard `alloc` that has access to the PSRAM would it work for the non DMA cases?

If I have in my code `with_hardware_rsa(&mut rsa)` does it mean I'm using crypto accelerator? I don't think I need performance so could probably remove that in favor of freeing up memory for static.

Also, it looks like the calls (from examples, and what I use) to `esp_alloc::heap_allocator` and `init_psram_heap` which ends in `Heap.add_region` go to the same allocator and just end up with two regions.
So when doing an allocation, how does esp-wifi gurantee the allocation would go to internal ram and not PSRAM? And the other way around, how does a standard allocation not taking space from esp-wifi pool?

**Reply by yanshay** [reply]
I tried overriding all the calloc/free/malloc/realloc that exist in esp-wifi, assuming they are there only for `esp-mbedtls`.
However, it seems like these are used also from `esp-wifi`/`esp-wifi-sys` so the application hangs as soon as it reaches esp-wifi initialization.

Is there a way to have only `esp-mbedtls` use my psram allocation functions and other places to link with the esp-wifi allocation functions? I couldn't find a way to do such conditional linking.

I'm out of memory for my application and can't add anything now, and this would free up quite a lot of memory to give me more space. It would also make it usable on esp32 where even a basic example doesn't run.

**Reply by bjoernQ** [reply]
`calloc` is used a lot by the wifi drivers - I _think_ you would need to build mbedtls C code yourself and set the right things in the configuration header file to do that

BTW are you using `dram2` region as additional heap? It's internal RAM which is not useable for static allocations or the heap: https://github.com/esp-rs/esp-hal/blob/c6404fe569519d1ac3c5ad865f41c077baba53e2/examples/src/bin/wifi_coex.rs#L62-L80

**Reply by yanshay** [reply]
I'm not (does it exist also on esp32s3?), but would it help?

I have lots of PSRAM for heap and using that for my allocations, my issue is that the stack hits the stuff that doesn't go on the alloc heap - esp-wifi/esp-mbedtls/embassy-tasks and my statics.

If I understand correctly the dram2 cant be used for those.

P.S. if there are multiple regions to the heap, how does alloc decides where to allocate first? Could it be my allocations eat out of the esp-wifi heap first and that's why I need to have it so large (close to 200k with esp-mbedtls)?

**Reply by bjoernQ** [reply, marked as answer]
Yes it's also available on S3 now (since esp-hal 0.21.0) - it's internal RAM which is used by the bootloader during boot.

You could reduce the allocation area in `dram1` and have more space for the stack then

`esp-alloc` will try the first region which qualifies for the required type of memory and tries the next region if it fails.
"Normal" allocations don't require anything special which means they will first "eat up"  the first added region first

**Reply by yanshay** [reply]
So basically what I should do is:
1. Initialize PSRAM region heap first so any standard allocation will end up there
2. Reduce the current internal ram alloc I have (which conflicts with stack) by 64k
3. Add the dram2 alloc of 64k - esp-wifi can use it as well.

This will increase the size available to the stack by 64k from my current situation.

Correct?

**Reply by bjoernQ** [reply]
Yes that should help I guess

**Reply by yanshay** [reply]
This worked great!
Even just switching order and adding psram first saved me quite some memory. And adding the extra 64K saved even more so I should. be good for quite some time now. Thanks!

I don't think I've seen that information mentioned anywhere, so worth documenting somewhere. It's one of those things that cause crashes that are difficult to track, and this would avoid it altogether.

Some more thoughts around this:
1. Worth adding a macro like there is `heap_allocator!` called `dram2_heap_allocator!` or something like that make sense.
Maybe even if possible automate it so that `esp-alloc` would split the heap reservation to first those dram2 heap and the rest on the standard memory, probably doable with macros.
2. Worth considerin having by default `esp-alloc` try to reserve always first in PSRAM if region added. If someone added that region it's because they have memory issues and would want everything to go there if possible. I can see cases where that wouldn't be what the user wants (to maximize memory speed) but as default might make sense.
